examples/jsm/loaders/KTX2Loader.js

```diff
@@ -59,224 +66,494 @@ class KTX2Loader extends CompressedTextureLoader {
 
 	setTranscoderPath( path ) {
 
-		this.basisLoader.setTranscoderPath( path );
+		this.transcoderPath = path;
 
 		return this;
 
 	}
 
-	setWorkerLimit( path ) {
+	setWorkerLimit( num ) {
 
-		this.basisLoader.setWorkerLimit( path );
+		this.workerPool.setWorkerLimit( num );
 
 		return this;
 
 	}
 
 	detectSupport( renderer ) {
 
-		this.basisLoader.detectSupport( renderer );
+		this.workerConfig = {
+			astcSupported: renderer.extensions.has( 'WEBGL_compressed_texture_astc' ),
+			etc1Supported: renderer.extensions.has( 'WEBGL_compressed_texture_etc1' ),
+			etc2Supported: renderer.extensions.has( 'WEBGL_compressed_texture_etc' ),
+			dxtSupported: renderer.extensions.has( 'WEBGL_compressed_texture_s3tc' ),
+			bptcSupported: renderer.extensions.has( 'EXT_texture_compression_bptc' ),
+			pvrtcSupported: renderer.extensions.has( 'WEBGL_compressed_texture_pvrtc' )
+				|| renderer.extensions.has( 'WEBKIT_WEBGL_compressed_texture_pvrtc' )
+		};
 
 		return this;
 
 	}
 
 	dispose() {
 
-		this.basisLoader.dispose();
+		this.workerPool.dispose();
+		if ( this.workerSourceURL ) URL.revokeObjectURL( this.workerSourceURL );
 
 		return this;
 
 	}
 
+	init() {
+
+		if ( ! this.transcoderPending ) {
+
+			// Load transcoder wrapper.
+			const jsLoader = new FileLoader( this.manager );
+			jsLoader.setPath( this.transcoderPath );
+			jsLoader.setWithCredentials( this.withCredentials );
+			const jsContent = jsLoader.loadAsync( 'basis_transcoder.js' );
+
+			// Load transcoder WASM binary.
+			const binaryLoader = new FileLoader( this.manager );
+			binaryLoader.setPath( this.transcoderPath );
+			binaryLoader.setResponseType( 'arraybuffer' );
+			binaryLoader.setWithCredentials( this.withCredentials );
+			const binaryContent = binaryLoader.loadAsync( 'basis_transcoder.wasm' )
+
+			this.transcoderPending = Promise.all( [ jsContent, binaryContent ] )
+				.then( ( [ jsContent, binaryContent ] ) => {
+
+					const fn = KTX2Loader.BasisWorker.toString();
+
+					const body = [
+						'/* constants */',
+						'let _EngineFormat = ' + JSON.stringify( KTX2Loader.EngineFormat ),
+						'let _TranscoderFormat = ' + JSON.stringify( KTX2Loader.TranscoderFormat ),
+						'let _BasisFormat = ' + JSON.stringify( KTX2Loader.BasisFormat ),
+						'/* basis_transcoder.js */',
+						jsContent,
+						'/* worker */',
+						fn.substring( fn.indexOf( '{' ) + 1, fn.lastIndexOf( '}' ) )
+					].join( '\n' );
+
+					this.workerSourceURL = URL.createObjectURL( new Blob( [ body ] ) );
+					this.transcoderBinary = binaryContent;
+
+					this.workerPool.setWorkerCreator( () => {
+
+						const worker = new Worker( this.workerSourceURL );
+						const transcoderBinary = this.transcoderBinary.slice( 0 );
```

Comment:
why do you explicitly copy this arraybuffer and then transfer it?
it makes no sense, just let postMessage copy it the optimal way.

Reply:
We're explicit rather than implicit about the copy because it's required — we'll need the binary on the main thread again later. Any reason to expect `.slice()` is less efficient than the copy postMessage would do?

Reply:
remove the binary from postMessage's transfer list (2nd parameter) and it will live on the main thread just fine.
`.slice()` always makes a copy, but postMessage can avoid it in certain cases.
for example: if buffer is immutable, it will be a zero-copy operation - just increment the refcount.

Reply:
well, after investigating the implementations of some javascript VMs, I am withdrawing my comments.
looks like current slice & transfer method is really the most optimal one for this ArrayBuffer case (not for SharedArrayBuffer though).

what happens in some VMs when slice & transfer is *not* used:
1. ArrayBuffer is serialized to message (first copy)
2. message delivered
3. new ArrayBuffer is deserialized from message (second copy)

what happens in all VMs when slice & transfer is used:
1. new ArrayBuffer is created by slice (copy)
2. small ArrayBuffer handle is serialized to message
3. message delivered
4. small ArrayBuffer handle is deserialized from message

So, there is a guarantee of no more than 1 whole buffer copy when slice & transfer is used.